flowzone.yml

```diff
@@ -2980,6 +3045,161 @@ jobs:
+          caps="$(echo "${stack}" | jq -rc .capabilities[] | paste -sd' ' -)"
+
+          echo "stack_name=${{ matrix.stack }}" >> $GITHUB_OUTPUT
+          echo "template_file=${template_file}" >> $GITHUB_OUTPUT
+          echo "tags=${tags}" >> $GITHUB_OUTPUT
+          echo "params=${params}" >> $GITHUB_OUTPUT
+          echo "kvparams=${kvparams}" >> $GITHUB_OUTPUT
+          echo "caps=${caps}" >> $GITHUB_OUTPUT
+
+      - name: Validate template
+        run: |
+          aws cloudformation validate-template \
+            --template-body 'file://${{ steps.shared.outputs.template_file }}'
+
+      - name: Package template
+        run: |
+          mkdir -p "package/$(dirname '${{ steps.shared.outputs.template_file }}')"
+
+          aws cloudformation package \
+            --template-file '${{ steps.shared.outputs.template_file }}' \
+            --s3-bucket '${{ steps.make_bucket.outputs.s3_bucket }}' \
+            --output-template-file 'package/${{ steps.shared.outputs.template_file }}'
+
+      - name: Estimate costs
+        run: |
+          aws cloudformation estimate-template-cost \
+            --template-body 'file://package/${{ steps.shared.outputs.template_file }}' \
+            --parameters ${{ steps.shared.outputs.kvparams }}
+
+      - name: Generate change set
+        id: change_set
+        run: |
+          result="$(aws cloudformation deploy \
+            --stack-name '${{ steps.shared.outputs.stack_name }}' \
+            --template-file 'package/${{ steps.shared.outputs.template_file }}' \
+            --s3-bucket '${{ steps.make_bucket.outputs.s3_bucket }}' \
+            --capabilities ${{ steps.shared.outputs.caps }} \
+            --parameter-overrides '${{ steps.shared.outputs.params }}' \
+            --tags ${{ steps.shared.outputs.tags }} \
+            --no-fail-on-empty-changeset \
+            --no-execute-changeset)"
+
+          if ! [[ "$result" =~ 'No changes to deploy' ]]; then
+              cmd="${result#*:}"
+              cmd=${cmd//$'\n'/}
+              echo "command=${cmd}" >> $GITHUB_OUTPUT
+          else
+              echo '::info::no changes'
+          fi
+
+      - name: Describe change set(s)
+        if: steps.change_set.outputs.command != ''
+        run: |
+          result="$(${{ steps.change_set.outputs.command }})"
+          if grep -qEi 'replace|delete'; then
+              echo '::warning::change set may destroy existing resources'
+          fi
+          echo "${result}" | jq -r
+
+  cloudformation_finalize:
+    name: Finalize CloudFormation
+    runs-on: ${{ fromJSON(inputs.runs_on) }}
+    strategy:
+      fail-fast: false
+      matrix:
+        stack: ${{ fromJSON(needs.is_cloudformation.outputs.stacks) }}
+    timeout-minutes: ${{ fromJSON(inputs.jobs_timeout_minutes) }}
+    needs:
+      - is_pr_merged
+      - is_pushed_tag
+      - is_cloudformation
+      - cloudformation_test
```

Comment:
`cloudformation_test` only runs on open PRs so this finalize job will never run.

If test and finalize need to share identical steps, those steps should be made into yaml anchors, because they never run as part of the same workflow.